distributed/shuffle/_worker_extension.py

```diff
@@ -185,15 +200,22 @@ def heartbeat(self) -> dict[str, Any]:
             "start": self.start_time,
         }
 
-    async def receive(self, data: list[bytes]) -> None:
+    async def receive(self, data: list[tuple[int, bytes]]) -> None:
         await self._receive(data)
 
-    async def _receive(self, data: list[bytes]) -> None:
+    async def _receive(self, data: list[tuple[int, bytes]]) -> None:
         self.raise_if_closed()
 
+        filtered = []
+        for d in data:
+            if d[0] not in self.received:
+                filtered.append(d[1])
+                self.received.add(d[0])
+                self.total_recvd += sizeof(d)
+        if not filtered:
```

Comment:
I suggest to `del data` here to clear references to the data that was not received